SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -1935,29 +1934,44 @@ void USpatialReceiver::ApplyComponentUpdate(const Worker_ComponentUpdate& Compon
 	}
 }
 
-ERPCResult USpatialReceiver::ApplyRPCInternal(UObject* TargetObject, UFunction* Function, const RPCPayload& Payload, const FString& SenderWorkerId, bool bApplyWithUnresolvedRefs /* = false */)
+FRPCErrorInfo USpatialReceiver::ApplyRPCInternal(UObject* TargetObject, UFunction* Function, const FPendingRPCParams& Params, const FString& SenderWorkerId, bool bApplyWithUnresolvedRefs /* = false */)
 {
-	ERPCResult Result = ERPCResult::Unknown;
+	FRPCErrorInfo ErrorInfo = { TargetObject, Function };
 
 	uint8* Parms = (uint8*)FMemory_Alloca(Function->ParmsSize);
 	FMemory::Memzero(Parms, Function->ParmsSize);
 
 	TSet<FUnrealObjectRef> UnresolvedRefs;
 	TSet<FUnrealObjectRef> MappedRefs;
-	RPCPayload PayloadCopy = Payload;
+	RPCPayload PayloadCopy = Params.Payload;
 	FSpatialNetBitReader PayloadReader(PackageMap, PayloadCopy.PayloadData.GetData(), PayloadCopy.CountDataBits(), MappedRefs, UnresolvedRefs);
 
 	TSharedPtr<FRepLayout> RepLayout = NetDriver->GetFunctionRepLayout(Function);
 	RepLayout_ReceivePropertiesForRPC(*RepLayout, PayloadReader, Parms);
 
-	if ((UnresolvedRefs.Num() == 0) || bApplyWithUnresolvedRefs)
+	if (UnresolvedRefs.Num() != 0 && !bApplyWithUnresolvedRefs)
 	{
-		TargetObject->ProcessEvent(Function, Parms);
-		Result = ERPCResult::Success;
+		ErrorInfo.ErrorCode = ERPCResult::UnresolvedParameters;
 	}
 	else
 	{
-		Result = ERPCResult::UnresolvedParameters;
+		// Get the RPC target Actor.
+		AActor* Actor = TargetObject->GetTypedOuter<AActor>();
+
+		const bool bHasServerEndpointAuthority = StaticComponentView->HasAuthority(Params.ObjectRef.Entity, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID);
+		if (bHasServerEndpointAuthority &&
+			Actor->Role == ROLE_SimulatedProxy &&
+			Params.Type != ERPCType::NetMulticast)
```

Comment:
I believe we need this condition instead. We want to catch cases where the references got resolved (or timed out) when the server knows it should no longer have authority (so it set the role to `SimulatedProxy`), but it may or may not still have authority on the server endpoint.
```suggestion
		if ((Params.Type == ERPCType::ServerReliable ||
			Params.Type == ERPCType::ServerUnreliable) &&
			Actor->Role == ROLE_SimulatedProxy)
```